### ISSUE TITLE
extremely poor performance on inventory source deletion

### ISSUE THREAD

**Issue by wenottingham** [ansible/awx#4485]
##### ISSUE TYPE
 - Bug Report

##### SUMMARY
1. create an inventory source
2. hook it up to an AWS account that will return ~2400 or so instances
3. delete the source

This causes AWX to send three DELETE requests - to N/hosts, N/groups, and N:
```
/api/v2/inventory_sources/N/hosts:
X-API-Time | 43.693s
X-API-Total-Time | 57.284s
```
```
/api/v2/inventory_sources/N/groups:
504, killed due to request/response timeout, I assume:
Wed Aug 14 18:23:36 2019 - *** HARAKIRI ON WORKER 2 (pid: 741, try: 1) ***
Wed Aug 14 18:23:36 2019 - HARAKIRI !!! worker 2 status !!!
Wed Aug 14 18:23:36 2019 - HARAKIRI [core 0] 192.168.124.1 - DELETE /api/v2/inventory_sources/33/groups/ since 1565806895
Wed Aug 14 18:23:36 2019 - HARAKIRI !!! end of worker 2 status !!!
```

##### ENVIRONMENT
* AWX version: current and prior

**Comment by AlanCoding** [comment]
Prior related fix https://github.com/ansible/awx/pull/3412

That could have been extended to bulk delete the hosts/groups to address this issue, but we cannot, because there is still a mode where activity stream is recorded for deletion of each individual host or group.

**Comment by AlanCoding** [comment]
https://github.com/ansible/awx/blob/08df2cad683ca2375ca53d484b6d5d4a8123035a/awx/main/conf.py#L29

https://github.com/ansible/awx/blob/4c697ae477d1043098746b89b9f78861abec43d6/awx/settings/defaults.py#L952

**Comment by AlanCoding** [comment]
I stand corrected, the default behavior is that activity stream entries are recorded for all the deletions, irrelevant of that setting. Clearly, that's what the hold up is.

**Comment by elyezer** [comment]
I've synced an AWS account and got the following info message `Loaded 750 groups, 437 hosts`

Deleting the source I got the following time for deleting the groups:

```
X-API-Time: 6.017s
X-API-Total-Time: 6.164s
```

And the following time for deleting the hosts:

```
X-API-Time: 0.538s
X-API-Total-Time: 0.699s
```

Host deletion seems to be fine but group deleting seems to be having a poor performance still. I am trying to reproduce a much larger number of hosts and groups using inventory scripts and will get back with the results later.

**Comment by elyezer** [comment]
Using the following inventory script that generates one groups with 2000 hosts:

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import json
inventory = dict()
group_name = 'group_1'
inventory[group_name] = dict()
inventory[group_name]['hosts'] = [
    "host_{}".format(i)
    for i in range(2000)
]
inventory[group_name]['vars'] = dict(ansible_host='127.0.0.1', ansible_connection='local')
print(json.dumps(inventory))
```

Here are the times for deleting groups and hosts respectively:

```
# groups
X-API-Time: 0.144s
X-API-Total-Time: 0.195s

# hosts
X-API-Time: 0.675s
X-API-Total-Time: 1.015s
```

**Comment by elyezer** [comment]
Using the following inventory script that generates 1000 groups with one host on each:

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import json
inventory = dict()

for group_id in range(1000):
    inventory['group_{}'.format(group_id)] = {
        'hosts': ['host_{}'.format(group_id)],
        'vars': dict(ansible_host='127.0.0.1', ansible_connection='local')
    }
print(json.dumps(inventory))
```

Here are the times for deleting groups and hosts respectively:

```
# groups
X-API-Time: 4.620s
X-API-Total-Time: 4.795s

# hosts
X-API-Time: 0.999s
X-API-Total-Time: 1.325s
```

**Comment by elyezer** [comment]
It seems that deleting groups are kind of expensive task still. Do we want to take a better look on it @AlanCoding @wenottingham ?

**Comment by wenottingham** [comment]
What is the 'before' for these examples?

**Comment by AlanCoding** [comment]
Honestly, 6 seconds for 750/437 is markedly better than what I hoped for. I suspect that you'll see a higher penalty if you do them simultaneous, as the UI does. As much as twice that would match my own expectations.

My reference case was about 400 groups (same group-host ratio) from AWS, which was around the uWSGI timeout borderline. In that case, the timeout was flaky, sometimes it would happen, sometimes it wouldn't. But the timeout is 60 seconds.

**Comment by ryanpetrello** [comment]
@elyezer I'd be curious to see what the timing was if you rolled back @AlanCoding's change and compared with the same data.  I think @wenottingham was reporting things in the neighborhood of a minute+

**Comment by elyezer** [comment]
Since the script with more groups had the worst performance, I've used it to generate 1000 groups with one host on each:

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import json
inventory = dict()

for group_id in range(1000):
    inventory['group_{}'.format(group_id)] = {
        'hosts': ['host_{}'.format(group_id)],
        'vars': dict(ansible_host='127.0.0.1', ansible_connection='local')
    }
print(json.dumps(inventory))
```

Have patched the instance to rever changes added by #3817 and collected the times for deleting groups and hosts respectively:

```
# groups
X-API-Time: 14.510s
X-API-Total-Time: 18.084s

# hosts
X-API-Time: 31.357s
X-API-Total-Time: 34.338s
```

With that we can confirm that #3817 changes are making a huge performance improvement, specially on the hosts deletion, and we can close this one.
